src/wallet/rpcnames.cpp

```diff
@@ -676,6 +677,144 @@ name_update (const JSONRPCRequest& request)
+        "       \"name\"              (string, optional) Name to watch for confirmation\n"
+        "       \"confirmations\"     (numeric, optional) Queued transaction will be broadcast when the txid or name (most recent update) has this many confirmations\n"
+        "     }\n"
+        "   }\n"
+        "\nResult:\n"
+        "\"txid\"             (string) the queued transaction's hash in hex\n"
+        // TODO: Examples
+      );
+
+  RPCTypeCheck (request.params,
+                {UniValue::VSTR, UniValue::VOBJ});
+
+  // parse transaction from parameter
+  // we only use this for error checking and returning a txid
+  CMutableTransaction mtxParsed;
+  if (!DecodeHexTx(mtxParsed, request.params[0].get_str()))
+    throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+  CTransactionRef txParsed(MakeTransactionRef(std::move(mtxParsed)));
+  const uint256& hashTx = txParsed->GetHash();
+
+  const valtype tx
+      = DecodeNameFromRPCOrThrow (request.params[0], NameEncoding::HEX);
+
+  UniValue options(UniValue::VOBJ);
+  if (request.params.size () >= 2)
+    options = request.params[1].get_obj ();
+
+  RPCTypeCheckObj (options,
+    {
+      {"sendWhen", UniValueType (UniValue::VOBJ)},
+    },
+    true, false);
+
+  UniValue sendWhen(UniValue::VOBJ);
+  if (options.exists ("sendWhen"))
+    sendWhen = options["sendWhen"].get_obj ();
+
+  RPCTypeCheckObj (sendWhen,
+    {
+      {"txid", UniValueType (UniValue::VSTR)},
+      {"name", UniValueType (UniValue::VSTR)},
+      {"confirmations", UniValueType (UniValue::VNUM)},
+    },
+    true, false);
+
+  uint256 triggerTxid = uint256();
+  if (sendWhen.exists ("txid"))
+    triggerTxid = ParseHashV (sendWhen["txid"], "sendWhen txid");
+
+  valtype triggerName(0);
+  if (sendWhen.exists ("name"))
+    triggerName = DecodeNameFromRPCOrThrow (sendWhen["name"], ConfiguredNameEncoding ());
+
+  int32_t triggerDepth = 0;
+  if (sendWhen.exists ("confirmations"))
+    triggerDepth = sendWhen["confirmations"].get_int();
+
+  LOCK2 (cs_main, pwallet->cs_wallet);
+
+  if (!pwallet->WritePendingNameFirstUpdate(hashTx.GetHex(), tx, triggerTxid, triggerName, triggerDepth))
+  {
+    throw JSONRPCError (RPC_WALLET_ERROR, "Error queueing transaction");
+  }
+
+  return hashTx.GetHex();
+}
+
+/* ************************************************************************** */
+
+UniValue
+listqueuedtransactions (const JSONRPCRequest& request)
+{
+  std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest (request);
+  CWallet* const pwallet = wallet.get ();
+
+  if (!EnsureWalletIsAvailable (pwallet, request.fHelp))
+    return NullUniValue;
+
+  if (request.fHelp || request.params.size () > 0)
+    throw std::runtime_error (
+        "listqueuedtransactions\n"
+        "\nList the transactions that are queued for future broadcast.\n"
+      );
+
+  LOCK2 (cs_main, pwallet->cs_wallet);
+
+  UniValue result(UniValue::VOBJ);
+
+  for (const auto& i : wallet->namePendingMap)
```

Comment:
I don't really like that you use `namePendingMap` directly here.  Please add a function that returns the map, so that you can make the variable `private` in `CWallet`.

Reply:
Should the function return a copy of the map (thus avoiding the need for the caller to worry about concurrent access)?  I suspect that the current usage may not be thread-safe.

Reply:
FWIW it looks like upstream `CWallet` uses a public `std::map` (with `GUARDED_BY` macro for thread safety) for `mapWallet`.  Any reason not to just do what upstream does here?